File: backend/app_service/models/community_reply.py
"""CommunityReply model — maps to community_replies table (DATA_MODEL.md §11)."""

from datetime import datetime

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class CommunityReply(Base):
    """Community Q&A reply — soft-delete enabled."""

    __tablename__ = "community_replies"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)
    post_id: Mapped[str] = mapped_column(String(36), nullable=False)
    user_id: Mapped[str] = mapped_column(String(128), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    is_best_answer: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    upvote_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    ai_moderation_status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="pending"
    )
    ai_moderation_reason: Mapped[str | None] = mapped_column(Text, nullable=True, default=None)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )
    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, default=None
    )

File: app/lib/features/community/domain/community_reply.dart
import 'community_post.dart';

/// A reply to a community post.
class CommunityReply {
  const CommunityReply({
    required this.id,
    required this.postId,
    required this.userId,
    required this.content,
    required this.isBestAnswer,
    required this.upvoteCount,
    required this.moderationStatus,
    required this.userVoted,
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String postId;
  final String userId;
  final String content;
  final bool isBestAnswer;
  final int upvoteCount;
  final ModerationStatus moderationStatus;
  final bool userVoted;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  factory CommunityReply.fromJson(Map<String, dynamic> json) {
    return CommunityReply(
      id: json['id'] as String,
      postId: json['post_id'] as String? ?? '',
      userId: json['user_id'] as String? ?? '',
      content: json['content'] as String? ?? '',
      isBestAnswer: json['is_best_answer'] as bool? ?? false,
      upvoteCount: json['upvote_count'] as int? ?? 0,
      moderationStatus: ModerationStatus.fromString(
          json['ai_moderation_status'] as String? ?? 'pending'),
      userVoted: json['user_voted'] as bool? ?? false,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'] as String)
          : null,
    );
  }

  CommunityReply copyWith({
    String? id,
    String? postId,
    String? userId,
    String? content,
    bool? isBestAnswer,
    int? upvoteCount,
    ModerationStatus? moderationStatus,
    bool? userVoted,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return CommunityReply(
      id: id ?? this.id,
      postId: postId ?? this.postId,
      userId: userId ?? this.userId,
      content: content ?? this.content,
      isBestAnswer: isBestAnswer ?? this.isBestAnswer,
      upvoteCount: upvoteCount ?? this.upvoteCount,
      moderationStatus: moderationStatus ?? this.moderationStatus,
      userVoted: userVoted ?? this.userVoted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
